### ISSUE TITLE
Generic string literal parameter not enforced

### ISSUE THREAD

**Issue by GillesDebunne** [Microsoft/TypeScript#31097]
**TypeScript Version:**  3.4.5

**Search Terms:**
Generic, string literal type, invalid parameter

**Code**

```ts
interface Func<P> {
  (param: P): void;
}

interface AorB {
  value?: 'aaa' | 'bbb';
}

// Correct, no error
const f1: Func<AorB> = ({ value = 'aaa' }: AorB) => {
  console.log(value);
};

// Incorrect, no error is detected although it should be equivalent to f3 below
const f2: Func<AorB> = ({ value = 'invalid' }) => {
  console.log(value);
};

// Correct, error because `invalid` is not a valid value
const f3: Func<AorB> = ({ value = 'invalid' }: AorB) => {
  console.log(value);
};
```

**Expected behavior:**
Function `f1` correctly defines a default `aaa` for value.
Function `f2` defines an invalid default value, and triggers a compile error.
Function `f3` defines an invalid default value, and triggers a compile error. The difference with `f2` is that the `AorB` parameter type is repeated.

**Actual behavior:**
Only `f3` raises an error. For `f2`, the inferred type of `value` is `'aaa'|'bbb'|'invalid'`, which seems incorrect.

**Playground Link:**
https://www.typescriptlang.org/play/index.html#src=interface%20Func%3CP%3E%20%7B%0D%0A%20%20(param%3A%20P)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20AorB%20%7B%0D%0A%20%20value%3F%3A%20'aaa'%20%7C%20'bbb'%3B%0D%0A%7D%0D%0A%0D%0Aconst%20f1%3A%20Func%3CAorB%3E%20%3D%20(%7B%20value%20%3D%20'aaa'%20%7D%3A%20AorB)%20%3D%3E%20%7B%0D%0A%20%20console.log(value)%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20f2%3A%20Func%3CAorB%3E%20%3D%20(%7B%20value%20%3D%20'invalid'%20%7D)%20%3D%3E%20%7B%0D%0A%20%20console.log(value)%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20f3%3A%20Func%3CAorB%3E%20%3D%20(%7B%20value%20%3D%20'invalid'%20%7D%3A%20AorB)%20%3D%3E%20%7B%0D%0A%20%20console.log(value)%3B%0D%0A%7D%3B%0D%0A

**Related Issues:**
No related open issues found with terms generic and string literal

**Comment by jcalz** [comment]
Looks like `f2` compiles because a function that accepts a parameter of type `{value: 'aaa'|'bbb'|'invalid'}` is also a function that accepts a parameter of type `{value: 'aaa'|'bbb'}`.  That is, it's perfectly fine to assign a function that accepts a wider parameter to a function type expecting a narrower parameter (usually we say that functions are [contravariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Contravariant_method_argument_type) in their parameters types).  

The reason there's an error in the `f3` line is because you are saying that a default value of type `{value: 'invalid'}` is assignable to `AorB`, which it's not.  This has nothing to do with assigning a function to a variable of type `Func<AorB>`.  In fact you can remove the assignment and it's still an error:
```ts
({ value = 'invalid' }: AorB) => { // still error
  console.log(value);
};
```

So I'm thinking that this will be closed as "working as intended", unless there's something big I'm missing here.

**Comment by GillesDebunne** [comment]
Thanks for the comment. It makes sense, although it feels weird to extend a type with a default value.
